src/lxc/lxccontainer.c

```diff
@@ -2320,7 +2322,12 @@ static bool remove_from_array(char ***names, char *cname, int size)
 {
 	char **result = get_from_array(names, cname, size);
 	if (result != NULL) {
-		free(result);
+		int i;
+		for (i = 0; (*names)[i] != *result && i < size; i++) {
+		}
+		free(*result);
+		memmove(*names+i, *names+i+1, (size-i-1) * sizeof(char*));
+		*names = (char**)realloc(*names, (size-1) * sizeof(char *));
```

Comment:
Same as above in `add_to_array`.

Reply:
I would disagree here. If you leave it without shrinking, then size of an array becomes bigger than the number of elements the array has. It brings the following issues:
- When iterating over the array you access to elements that are already freed. Such iteration leads to duoble-free at the end of every function that has called remove_from_array(), and to crash. The simple NULLing the element via `*result = NULL` after `free()` does not help too much because of the second issue.
- When adding a new element after one was freed, the call to realloc() causes memory leak, because the number of elements (`pos+1`) is passed to realloc() instead of the real size of the array:
  ```
  char **newnames = (char**)realloc(*names, (pos+1) * sizeof(char *));
  ```

Please, refer to this output produced by Valgring:
```
$ clang++ -std=c++11 -g -O0 -Wall -Wno-disabled-macro-expansion -Wno-writable-strings -Wno-deprecated lxc-top.test.c -o lxc-top.test.out
$ valgrind ./lxc-top.test.out

==1690443== Memcheck, a memory error detector
==1690443== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==1690443== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==1690443== Command: ./lxc-top.test.out
==1690443==
i[0]='Item#01'
i[1]='Item#02'
i[2]='Item#03'
i[3]='Item#04'
i[0]='Item#01'
i[1]='Item#02'
i[2]='(null)'
==1690443==
==1690443== HEAP SUMMARY:
==1690443==     in use at exit: 8 bytes in 1 blocks
==1690443==   total heap usage: 10 allocs, 9 frees, 73,840 bytes allocated
==1690443==
==1690443== LEAK SUMMARY:
==1690443==    definitely lost: 8 bytes in 1 blocks
==1690443==    indirectly lost: 0 bytes in 0 blocks
==1690443==      possibly lost: 0 bytes in 0 blocks
==1690443==    still reachable: 0 bytes in 0 blocks
==1690443==         suppressed: 0 bytes in 0 blocks
==1690443== Rerun with --leak-check=full to see details of leaked memory
==1690443==
==1690443== For lists of detected and suppressed errors, rerun with: -s
==1690443== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

And the code:
```c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define ERROR(x) do { fprintf(stderr, x); fprintf(stderr, "\n"); } while (0)

static inline int string_cmp(char **first, char **second)
{
	return strcmp(*first, *second);
}

static char** get_from_array(char ***names, char *cname, int size)
{
	return (char **)bsearch(&cname, *names, size, sizeof(char *), (int (*)(const void *, const void *))string_cmp);
}

static bool add_to_array(char ***names, char *cname, int pos)
{
	char **newnames = (char**)realloc(*names, (pos+1) * sizeof(char *));
	if (!newnames) {
		ERROR("Out of memory");
		return false;
	}

	*names = newnames;
	newnames[pos] = strdup(cname);
	if (!newnames[pos])
		return false;

	// sort the arrray as we will use binary search on it
	qsort(newnames, pos + 1, sizeof(char *), (int (*)(const void *,const void *))string_cmp);

	return true;
}

static bool remove_from_array(char ***names, char *cname, int size)
{
	char **result = get_from_array(names, cname, size);
	if (result != NULL) {
		free(*result);
		*result = NULL;
		return true;
	}
	return false;
}

void test_realloc()
{
	char **ct_name = NULL;
	int i, ct_name_cnt = 0;
	add_to_array(&ct_name, "Item#01", ct_name_cnt++);
	add_to_array(&ct_name, "Item#02", ct_name_cnt++);
	add_to_array(&ct_name, "Item#03", ct_name_cnt++);
	add_to_array(&ct_name, "Item#04", ct_name_cnt++);
	for (i = 0; i < ct_name_cnt; i++) {
		printf("i[%d]='%s'\n", i, ct_name[i]);
	}
	remove_from_array(&ct_name, "Item#03", ct_name_cnt--);
	for (i = 0; i < ct_name_cnt; i++) {
		printf("i[%d]='%s'\n", i, ct_name[i]);
	}
	for (i = 0; i < ct_name_cnt; i++) {
		free(ct_name[i]);
	}
	free(ct_name);
}

int main()
{
	test_realloc();
	return 0;
}

```